core/base/src/TROOT.cxx

```diff
@@ -2158,46 +2157,40 @@ TClass *TROOT::LoadClass(const char *requestedname, Bool_t silent) const
 Int_t TROOT::LoadClass(const char * /*classname*/, const char *libname,
                        Bool_t check)
 {
-   Int_t err = -1;
+   TString lib(libname);
 
-   char *path;
-   TString lib = libname;
-   if (!lib.BeginsWith("lib"))
-      lib = "lib" + lib;
-   if ((path = gSystem->DynamicPathName(lib, kTRUE))) {
+   // Check if libname exists in path or not
+   if (char *path = gSystem->DynamicPathName(lib, kTRUE)) {
+      // If check == true, only check if it exists and if it's readable
       if (check)
-         err = 0;
+         return 0;
+
+      // If check == false, try to load the library
       else {
-         err = gSystem->Load(path, 0, kTRUE);
+         int err = gSystem->Load(path, 0, kTRUE);
+         // TSystem::Load returns 1 when the library was already loaded, return success in this case.
+         if (err == 1)
+            err = 0;
+         return err;
```

Comment:
Each return needs to be prefixed with delete [] path; otherwise we have a memory leak.